sdk/core/Azure.Core.TestFramework/src/TestEnvironment.cs

```diff
@@ -26,6 +26,7 @@ public abstract class TestEnvironment
         [EditorBrowsableAttribute(EditorBrowsableState.Never)]
         public static string RepositoryRoot { get; }
 
+        private static readonly SemaphoreSlim s_environmentStateCacheSemaphore = new SemaphoreSlim(1, 1);
         private static readonly ConcurrentDictionary<Type, Task> s_environmentStateCache = new ConcurrentDictionary<Type, Task>();
```

Comment:
Make the dictionary non-concurrent.

Reply:
Is this safe? non-concurrent dictionary shouldn't be read if there's modification going on. The logic in the PR doesn't prevent that.
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-5.0#thread-safety

Reply:
I think we can do a plain old lock + dictionary

Reply:
okay, worst case we'll go back to concurrent dictionary + lock if things go south.